frontend/pages/upload.vue

```diff
@@ -0,0 +1,407 @@
+    })),
+    holdings_date: {
+      time: holdingsDate.value.toISOString(),
+    },
+  }).catch(e => {
+    console.log('error starting upload', e, e.body)
+    if (e.body?.error_id) {
+      errorCode.value = e.body.error_id
+    } else {
+      errorCode.value = 'Unknown Error'
+    }
+    if (e.body?.message) {
+      errorMessage.value = e.body.message
+    } else {
+      errorMessage.value = 'An unexpected error occurred - please file a bug to help us fix it.'
+    }
+    setAllStates(FileStatus.Selected)
+  })
+  if (!startPortfolioUploadResp) {
+    isProcessing.value = false
+    return
+  }
+
+  setAllStates(FileStatus.Uploading)
+  const uploads = fileStatesWithDetail.value.map(async (fileState: FileStateDetail): Promise<void> => {
+    const { file, index } = fileState
+    const respItem = startPortfolioUploadResp.items.find((item) => item.file_name === file.name)
+    if (!respItem) {
+      throw new Error(`Could not find start item for file ${file.name} - something is probably wrong in the API bizlogic`)
+    }
+    fileStates.value[index].incompleteUploadId = respItem.incomplete_upload_id
+
+    await $axios({
+      method: 'PUT',
+      url: respItem.upload_url,
+      data: file,
+      headers: {
+        'Content-Type': file.type,
+        'x-ms-blob-type': 'BlockBlob',
+      },
+    }).then(() => {
+      fileStates.value[index].status = FileStatus.Uploaded
+    }).catch((e) => {
+      console.log('error uploading file', e)
+      fileStates.value[index].status = FileStatus.Error
+      fileStates.value[index].errorMessage = serializeError(e)
+      errorCode.value = 'Upload Failed'
+      errorMessage.value = 'One or more files could not be uploaded - please delete/resolve them and try again.'
+    })
+  })
+  await Promise.all(uploads)
+
+  let hadError = false
+  for (const fileState of fileStates.value) {
+    if (fileState.status === FileStatus.Uploaded) {
+      continue
+    } else if (fileState.status === FileStatus.Error) {
+      hadError = true
+    } else {
+      throw new Error(`Unexpected file state ${fileState.status}`)
+    }
+  }
+  if (!hadError) {
+    await doParsing()
+  }
+  isProcessing.value = false
+}
+
+const doParsing = async () => {
+  fileStates.value.forEach((_, i) => {
+    fileStates.value[i].status = FileStatus.Validating
+  })
+  await pactaClient.completePortfolioUpload({
+    items: fileStates.value.map((fileState) => ({
+      incomplete_upload_id: presentOrFileBug(fileState.incompleteUploadId),
+    })),
+  })
+  await waitForValidationToCompleteOrTimeout()
+  await cleanUpIncompleteUploads()
+}
+
+const waitForValidationToCompleteOrTimeout = async () => {
+  const timeout = 1000 * 60 * 5
+  const start = Date.now()
+  do {
+    await refreshStateFromIncompleteUploads()
+    if (Date.now() - start > timeout) {
+      throw new Error('Timeout waiting for uploads to complete')
+    }
+    await new Promise((resolve) => setTimeout(resolve, 1000))
```

Comment:
[no action required]

I've never seen `await ... setTimeout` essentially used like `sleep` and it makes me feel weird, even if its totally reasonable.

Reply:
Nice! This is my (and it looks like the internet's) One liner:

https://stackoverflow.com/questions/37764665/how-to-implement-sleep-function-in-typescript